src/provider/dynamodb.ts

```diff
@@ -0,0 +1,198 @@
+'use strict';
+
+import { DynamoDB } from 'aws-sdk';
+import AWS = require('aws-sdk');
+import { DocumentClient, ItemList } from 'aws-sdk/clients/dynamodb';
+import { AWSConfig } from '../aws/config';
+import { Event } from '../model/event';
+import { Stream } from '../model/stream';
+import { PersistenceProvider } from './provider';
+
+
+/**
+ * A Persistence Provider that handle all the data in Dynamodb.
+ */
+export class DynamodbProvider implements PersistenceProvider {
+    private documentClient: DocumentClient;
+    private dynamoDB: DynamoDB;
+    private initialized = false;
+
+    constructor(awsConfig: AWSConfig) {
+        AWS.config.update(awsConfig);
+
+        this.documentClient = new DynamoDB.DocumentClient();
+        this.dynamoDB = new AWS.DynamoDB();
+    }
+
+    public async addEvent(stream: Stream, data: any) {
+        await this.ensureTables();
+
+        this.addAggregation(stream);
+        const commitTimestamp = Date.now();
+        const item = {
+            aggregation_streamid: `${this.getKey(stream)}`,
+            commitTimestamp: commitTimestamp,
+            payload: data,
+            stream: stream
+        };
+        const param = {
+            Item: item,
+            TableName: 'events',
+        };
+
+        await this.documentClient.put(param, (error, _) => {
+            if (error) {
+                throw new Error(error.message);
+            }
+        });
+
+        return {
+            commitTimestamp: commitTimestamp,
+            payload: data,
+        } as Event;
+    }
+
+
+    public async getEvents(stream: Stream, offset: number = 0, limit: number = -1): Promise<Array<Event>> {
+        await this.ensureTables();
+        const params = {
+            ConsistentRead: true,
+            ExpressionAttributeValues: {
+                ':a': this.getKey(stream)
+            },
+            KeyConditionExpression: 'aggregation_streamid = :a',
+            ScanIndexForward: false,
+            TableName: 'events',
+        };
+
+        const items: ItemList = (await this.documentClient.query(params).promise()).Items;
+
+        return items.map(data => {
+            return {
+                commitTimestamp: data.commitTimestamp,
+                payload: data.payload,
+            } as Event;
+        });
+    }
+
+    public async getAggregations(offset: number = 0, limit: number = -1): Promise<Array<string>> {
+        await this.ensureTables();
+        const params = {
+            TableName: 'aggregations',
```

Comment:
I am not sure how DynamoDB Works properly but instead of having a hard-coded `aggregations` couldn't we come up with a dynamic solution such as `tableName` = `stream+aggregation` etc?

Reply:
I didn't understand this suggestion, shall we talk?